iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/storable_function.inl

```diff
@@ -0,0 +1,343 @@
+{
+    if (src.empty())
+    {
+        // nothing to do, destroy and setting m_invoker to nullptr are performed before the type specific move
+        // operation
+        return;
+    }
+
+    auto ptr = dest.m_storage.template allocate<CallableType>();
+    if (ptr)
+    {
+        auto obj = reinterpret_cast<CallableType*>(src.m_callable);
+        ptr = new (ptr) CallableType(std::move(*obj));
+        dest.m_callable = ptr;
+        dest.m_invoker = src.m_invoker;
+        src.m_operations.destroy(src);
+        src.m_callable = nullptr;
+        src.m_invoker = nullptr;
+    }
+    else
+    {
+        // no memory in source, assignment could not be performed
+        // (this cannot happen in the static_storage case)
+        std::cerr << "storable_function: no memory to store moved object at destination\n";
+    }
+
+    cxx::Ensures(!dest.empty());
+}
+
+template <typename S, typename ReturnType, typename... Args>
+template <typename CallableType>
+void storable_function<S, signature<ReturnType, Args...>>::destroy(storable_function& f) noexcept
+{
+    // TODO: this is not an indication allocate was called anymore (due to fptrs also being stored in m_callable)
+    if (f.m_callable)
+    {
+        auto ptr = static_cast<CallableType*>(f.m_callable);
+        ptr->~CallableType();
+        f.m_storage.deallocate();
+    }
+}
+
+template <typename S, typename ReturnType, typename... Args>
+void storable_function<S, signature<ReturnType, Args...>>::copyFreeFunction(const storable_function& src,
+                                                                            storable_function& dest) noexcept
+{
+    dest.m_invoker = src.m_invoker;
+    dest.m_callable = src.m_callable;
+}
+
+template <typename S, typename ReturnType, typename... Args>
+void storable_function<S, signature<ReturnType, Args...>>::moveFreeFunction(storable_function& src,
+                                                                            storable_function& dest) noexcept
+{
+    dest.m_invoker = src.m_invoker;
+    dest.m_callable = src.m_callable;
+    src.m_invoker = nullptr;
+    src.m_callable = nullptr;
+}
+
+template <typename S, typename ReturnType, typename... Args>
+template <typename CallableType>
+ReturnType storable_function<S, signature<ReturnType, Args...>>::invoke(void* callable, Args&&... args)
+{
+    return (*static_cast<CallableType*>(callable))(std::forward<Args>(args)...);
+}
+
+template <typename S, typename ReturnType, typename... Args>
+ReturnType storable_function<S, signature<ReturnType, Args...>>::invokeFreeFunction(void* callable, Args&&... args)
+{
+    return (reinterpret_cast<ReturnType (*)(Args...)>(callable))(std::forward<Args>(args)...);
+}
+
+template <typename S, typename ReturnType, typename... Args>
+template <typename T>
+constexpr uint64_t storable_function<S, signature<ReturnType, Args...>>::required_storage_size() noexcept
+{
+    return sizeof(T) + alignof(T);
```

Comment:
First of all: I could save one 1 byte here and will do this, the pessimistic estimate is size + alignment - 1. 

The value the function here returns is supposed to be sufficient, not best possible (and it cannot be at compile time I think). Maybe we can do a little better, but I am not sure. `function` contains a `static_storage`, which is aligned somehow (Align).  This information can be used to improve on the estimate but it is a little more complicated since I think. Currently Align is set to 1 by default and used this way in function. For this, the estimate above is still best possible.

But `storable_function` has no knowledge about this alignment (and should not have, since the StorageType should be configurable).
TLDR: `storable_function` does not know `static_storage`, it is more generic.

Reply:
I introduced optimal allocation size calculation based on alignment, but at the price of adding a new requirement to the storage type of providing an `allocation_size` function (due to the reasons above).